Move special-case and tags from hard code to ui setting

benfoley opened the pull request:
This makes ui settings for the `special-cases` and `translation_tags` values that are used in cleaning the annotations.

It required added a ui_format property to the import settings config stuff so that more control could be had over using types of input components. The type of input the ui used was determined by the type of ui setting set, but this was limited to str, int and list. I wanted the UI for `special-cases` and `translation_tags` to be text area, and thought it best to have some mechanism for specifying that when defining the ui settings in elan transformer. 

Corresponding gui work is https://github.com/CoEDL/elpis-gui/pull/85

benfoley commented:
This moves

nicklambourne commented on elpis/transformer/__init__.py:
```diff
@@ -448,13 +448,14 @@ def import_setting(self, key, type, default=None, display_name=None, description
         self._import_ui_order_config.append(key)
     
-    def export_setting(self, key, type, default=None, display_name=None, description=None):
+    def export_setting(self, key, type, ui_format=None, default=None, display_name=None, description=None):
```
Can we get types on these parameters?

nicklambourne replied:
And make the comment params match the actual ones

nicklambourne commented on elpis/transformer/__init__.py:
```diff
@@ -498,7 +500,7 @@ def general_setting_title(self, title):
         return
 
-    def general_setting(self, key, type, default=None, display_name=None, description=None):
+    def general_setting(self, key, type, ui_format=None, default=None, display_name=None, description=None):
```
types and comment

nicklambourne commented on elpis/engines/common/objects/interface.py:
```diff
@@ -234,28 +232,23 @@ def list_models(self):
     def new_transcription(self, tname):
         if self.engine is None:
-            raise RuntimeError("Engine needs to be set prior to transcription")
+            raise RuntimeError("Engine need to be set prior to transcription")
```
This is a very small thing, but it seems to have gone from grammatically correct to not?

benfoley replied:
fixed

benfoley commented:
While adding types I also removed a setting value that was called "type". It was being used to do two things: first to set the type of UI element and also to hold info extracted from training files to build select options (eg with tier names etc). I removed this "type" value and added two separate parameters—ui_format and options. Now ui_format is used to determine the input component type, and options is used to hold file values and build the select options.

nicklambourne commented on elpis/transformer/__init__.py:
```diff
@@ -69,15 +66,15 @@ class DataTransformer:
+                 name: str,
+                 settings: Dict,
+                 ui: List,
+                 callback_get_config,
```
Should these have types, too?

benfoley replied:
Will have to ask Nic

nicklambourne commented on elpis/transformer/__init__.py:
```diff
@@ -419,133 +415,150 @@ def export(self, f: Callable):
-
-    def import_setting(self, key, type, ui_format=None, default=None, display_name=None, description=None):
+
+    def import_setting(self,
```
Missing types

benfoley replied:
arg, lost in the merge.. will add

benfoley replied:
added

nicklambourne commented on elpis/transformer/__init__.py:
```diff
@@ -419,133 +415,150 @@ def export(self, f: Callable):
         self._import_ui_order_config.append(key)
     
-    def export_setting(self, key, type, ui_format=None, default=None, display_name=None, description=None):
+    def export_setting(self,
```
Missing types

benfoley replied:
ditto

benfoley replied:
added

nicklambourne commented on elpis/transformer/__init__.py:
```diff
@@ -419,133 +415,150 @@ def export(self, f: Callable):
+                       resampled_path: str,
+                       temporary_directory_path: str,
+                       transcription_json_file_path: str,
+                       get_config_callback,
```
Missing types

nicklambourne commented on elpis/transformer/__init__.py:
```diff
@@ -741,19 +757,20 @@ def build_exporter(self,
+                  resampled_path: str,
+                  temporary_directory_path: str,
+                  transcription_json_file_path: str,
+                  get_config_callback,
```
If these are too hard to type, just mark all of these as resolved.

nicklambourne commented on elpis/transformer/__init__.py:
```diff
@@ -770,15 +787,16 @@ def make_importer(name: str,
+                  resampled_path: str,
+                  temporary_directory_path: str,
+                  transcription_json_file_path: str,
+                  get_config_callback,
```
And again
